python_modules/dagster/dagster/_core/definitions/assets/definition/asset_spec.py

```diff
@@ -308,6 +308,21 @@ def with_io_manager_key(self, io_manager_key: str) -> "AssetSpec":
             metadata={**self.metadata, SYSTEM_METADATA_KEY_IO_MANAGER_KEY: io_manager_key}
         )
 
+    def with_dagster_type(self, dagster_type: type) -> "AssetSpec":
+        """Returns a copy of this AssetSpec with an extra metadata value that dictates which
+        Dagster type to use to load the contents of this asset in downstream computations.
+
+        Args:
+            dagster_type (DagsterType): The Dagster type. This will be used as the value for the
+                "dagster/dagster_type" metadata key.
+
+        Returns:
+            AssetSpec
+        """
+        return self._replace(
+            metadata={**self.metadata, SYSTEM_METADATA_KEY_DAGSTER_TYPE: dagster_type}
```

Comment:
should we validate the dagster type here before putting it in metadata, just so the error message is more tractible?

Reply:
good call -- now this uses the `resolve_dagster_type` utility which does all sorts of validation (the set of things that can be resolved into DagsterTypes is not easily expressible in the type system)